lib/rucio/rse/rsemanager.py

```diff
@@ -317,7 +317,7 @@ def exists(rse_settings, files):
     files = [files] if not type(files) is list else files
     for f in files:
         exists = None
-        if (type(f) is str) or (type(f) is unicode):
+        if isinstance(f, str) or isinstance(f, type(u'')):
```

Comment:
Flake8 complained here, so I added this. seems OK. Not pretyt with the `type(u'')` though ...

Reply:
isinstance("asdfasdf", basestring)

tests for either string or unicode

Reply:
In py3 only str exists, in py2 str and unicode, hence (one of the) falke8 complaints. How about:

```python
if sys.version_info.major == 3:
    string_types = (str,)
else:
    string_types = (str, unicode)
if isinstance(f, string_types):
```

Reply:
Mind you flake still complains about this: `F821 undefined name 'unicode'`

Reply:
How about:
```python
if sys.version_info.major == 3:
    unicode = str
```

Reply:
So this is giving me a bad of a bad smell: I need to do this in each source file where we use the `unicode` keyword ... but I don't quite know how to do this as a module.

Reply:
So what I could do is make a `common/python3workaround.py` containing:
```
try:
    unicode
except NameError:
    unicode = str
```
then in any file which uses `unicode`
```
try:
    from rucio.common.python3workaround import unicode
except ImportError:
    pass
```
Long live exceptions?!

Reply:
we did something like that in some part of the code:
```
try:
    AUTHORIZED_VALUE_TYPES = (float, int, unicode)
except NameError:
    AUTHORIZED_VALUE_TYPES = (float, int, str)
```
generating the possible type values and then check against them

Reply:
if sys.version_info[0] == 2:
	strclass = basestring
elif sys.version_info[0] == 3:
	strclass = str
else:
	raise

Wondering why people in Python always try to replaces if's with exceptions.

Reply:
it's actually the recommended way. I read several times this advice.

Reply:
Only because a lot of people do it or it's marked as solution on stackoverflow it doesn't need to be recommended :D

Reply:
... mentioned with some motivation about the performance ;) but I doubt I can convince you anyway ;)

Reply:
@TWAtGH this is python [EAFP](https://docs.python.org/3/glossary.html#term-eafp) style/pattern.

Reply:
Yes I know but this code will only be executed once per start of a Rucio process. So this is no place were we should prefer performance over readability. Also EAFP should only be used when you know that a specific case is very exceptional and you don't know this in our case.

Reply:
At the moment, the rare case is running in python3. Plus the exception avoids adding the sys module.

Reply:
Not adding the sys module is a good point but I would guess that it's very unlikely that sys is not already loaded into memory, since it gets imported in some core areas of Rucio.
Again this is no point where performance matters because the impact is nearly nonexistent. Even if performance would matter the except cases should be really really rare.
The impact on the readability and maintainability would be huge though. One can immediately see that you're doing a differentiation between Python versions there. And think about adding more stuff that needs to be distinguished. With your approach you would end up adding a lot of try/excepts.

Reply:
Feature detection is already at various places of the code, following this:
https://docs.python.org/3/howto/pyporting.html#use-feature-detection-instead-of-version-detection

I would prefer to be consistent 

About unicode:
https://docs.python.org/3/howto/pyporting.html#text-versus-binary-data